### ISSUE TITLE
New linter for trailing assignments

### ISSUE THREAD

**Issue by ashbaldry** [r-lib/lintr#1491]
It would be nice to have a linter that flags whenever an assignment is at the end of the line. It can make function calls disjoined and harder to read.

Examples:
```r
y <- mean(x, na.rm =
  TRUE
)

some_long_variable_name <-
  function_name(...)
```

**Comment by IndrajeetPatil** [comment]
@ashbaldry Can you please check that the assignment that [tidyverse style guide](https://style.tidyverse.org/pipes.html#assignment-2) deems to be okay (the first example below) doesn't trigger a lint with this new linter? Also, please add test for the same.

<img width="815" alt="Screenshot 2022-07-31 at 13 51 38" src="https://user-images.githubusercontent.com/11330453/182024981-e275ac56-4ab2-4dd7-8709-1db94451ad39.png">

**Comment by ashbaldry** [comment]
Ah, I didn't see that section. Currently it will flag both the first and third situation. Let me add some tests in and see how to pass the situations above.

**Comment by ashbaldry** [comment]
@IndrajeetPatil I have included tests for all 3 versions and they are now all passing. Personally, I prefer to have the variable name and assignment on the same line so I have included a parameter that will flag the first and third situation, but by default it follows the tidyverse style. Hope that is alright, if not it is easy to remove.

**Comment by IndrajeetPatil** [comment]
Looks great, @ashbaldry! :) 

``` r
library(lintr)

lint(
  text = "x <- 5",
  linters = trailing_assignment_linter()
)

lint(
  text = "x <-
  5",
  linters = trailing_assignment_linter()
)
#> <text>:1:1: style: [trailing_assignment_linter] Assignment `<-` should not be trailing at end of line
#> x <-
#> ^

lint(
  text = "y <- mean(x, na.rm =
            TRUE
          )",
  linters = trailing_assignment_linter()
)
#> <text>:1:1: style: [trailing_assignment_linter] Assignment `=` should not be trailing at end of line
#> y <- mean(x, na.rm =
#> ^

lint(
  text = "some_long_variable_name <-
  function_name(...)",
  linters = trailing_assignment_linter()
)
#> <text>:1:1: style: [trailing_assignment_linter] Assignment `<-` should not be trailing at end of line
#> some_long_variable_name <-
#> ^

lint(
  text = "iris_long <-
  iris %>%
  gather(measure, value, -Species) %>%
  arrange(-value)"
  ,
  linters = trailing_assignment_linter()
)

lint(
  text = "iris_long <- iris %>%
  gather(measure, value, -Species) %>%
  arrange(-value)",
  linters = trailing_assignment_linter()
)

lint(
  text = "iris %>%
  gather(measure, value, -Species) %>%
  arrange(-value) ->
  iris_long",
  linters = trailing_assignment_linter()
)
```

<sup>Created on 2022-07-31 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

I think it looks good to me. Let's see if @MichaelChirico has any other concerns.

**Comment by AshesITR** [comment]
The third case is also triggered by `assignment_linter()`, so I suggest to extend the existing assignment linter by adding `allow_trailing = TRUE` to the mix, which would, if set to `FALSE` lint trailing assignment operators.
Otherwise, case 3 will produce two lints at the same location when using `linters_with_defaults(trailing_assignment_linter())`.

**Comment by MichaelChirico** [comment]
> I suggest to extend the existing assignment linter by adding allow_trailing = TRUE

oh, great idea. my biggest hesitation here was about adding a new linter for something I'm not sure has a v. broad appeal. which opens the can of worms of what criteria we want to use to assess new linters for non-tidyverse rules.

much easier to agree to a new argument to an existing linter.

**Comment by AshesITR** [comment]
Sorry for being a bit unclear there. I think it's still valuable to look at all assignments regarding line breaks.
However, `assignment_linter()` disallows all usages of `->` by default IIRC, so we only need to look for line breaks following any allowed assignment operator to avoid duplicate lints.
